channels/generic/multiplexer.py

```diff
@@ -0,0 +1,322 @@
+                {
+                    "type": "websocket.receive",
+                    "text": await self.encode_json(payload)
+                }
+            )
+            return
+
+        else:
+            raise ValueError("Invalid multiplexed **frame received (no channel/payload key)")
+
+    # Methods to INTERCEPT Stream Applications -> Client
+
+    async def websocket_send(self, message: Dict[str, Any], stream_name: str):
+        """
+        Capture downstream websocket.send messages from the stream applications.
+        """
+        text = message.get("text")
+        # todo what to do on binary!
+
+        json = await self.decode_json(text)
+        data = {
+            "stream": stream_name,
+            "payload": json
+        }
+
+        await self.send_json(data)
+
+    async def websocket_accept(self, message: Dict[str, Any], stream_name: str):
+        """
+        Intercept streams accepting the websocket connection and add them to
+        the set of accepted upstream streams.
+
+        An upstream Application must accept before messages will be
+        forwarded onto it.
+        """
+        self.accepted_upstream_stream.add(stream_name)
+
+    async def websocket_close(self, message: Dict[str, Any], stream_name: str):
+        """
+        Handle the downstream websocket.close message from a stream
+        application.
+        """
+        # just remove the stream from the set of open streams
+        self.accepted_upstream_stream.remove(stream_name)
+        # we don't cancel the task here, that will happen later.
+        # (when the main connection is closed).
+        # the application task could still send messages
+        # if it sends `websocket.accept` it will also start to recived messages
+        # again.
+        await self.close(code=message.get("code", None))
+
+    # De-multiplexing methods
+
+    async def __call__(self, receive, send):
+        """
+        Set up applications for each stream and set up self.
+        """
+        self.stream_upstream_queues, stream_tasks = await self._init_stream_applications()
+
+        # use gather here since we want to die as soon as any one raises
+        # an exception.
+
+        self._task_monitors = asyncio.gather(
+            *[self._monitor_task(name, task) for name, task in
+             stream_tasks.items()]
+        )
+
+        try:
+            # TODO this results in `Task was destroyed but it is pending!` sometimes
+
+            await asyncio.gather(
+                super().__call__(receive=receive, send=send),
+                self._task_monitors
+            )
+
+        finally:
+            try:
+                self._task_monitors.cancel()
+            finally:
+                # this might have raised an exception.
+                # cleanup stream applications
+                for task in stream_tasks.values():
+                    task.cancel()
+
+    async def _monitor_task(self, stream_name: str, task: asyncio.Task):
```

Comment:
StopConsumer should not bubble up out of other consumers (their coroutines will catch it), so I don't know why you have this method? Did you see it happen?

Reply:
don't we still need this method so that if just one child application dies (without an exception) we remove it from the set of upstream streams but keep the rest running.  

if we just gather/waiton the coroutines without some wrapper method we will not be able to remove them from the set of open upstream.

Reply:
don't we still need this method so that if just one child application dies (without an exception) we remove it from the set of upstream streams but keep the rest running?  

if we just gather/wait on the coroutines without some wrapper method we will not be able to remove them from the set of open upstream.